File: study/src/Patterns/proxy/Mine.java
package Patterns.proxy;

public class Mine extends Planet implements Runnable {
    private double warehouseCapacity = 1;
    private double warehouseLoad;
    private long timeForOnePiece = 1000;
    private double maxBatchSize = 2;

    public Mine(int x, int y, int radius, int r, int g, int b) {
        super(x, y, radius, r, g, b);
    }

    @Override
    public void run() {
        while (!Thread.currentThread().isInterrupted()) {
            if (warehouseLoad < warehouseCapacity) {
                warehouseLoad = warehouseLoad + 1;
                System.out.println("Warehouse load: " + warehouseLoad);
            }

            if (dockedShip != null) {
                double batchSize = maxBatchSize;
                if(dockedShip.getFreeCapacity() < maxBatchSize) {
                    batchSize = dockedShip.getFreeCapacity();
                }
                try {
                    System.out.println("Docked ship free capacity: " + dockedShip.getFreeCapacity());
                    System.out.println("Batch size: " + batchSize);
                    System.out.println("Ship load: " + dockedShip.getLoad());
                    dockedShip.batchLoad(batchSize);

                } catch (ShipOverloadedException e) {
                    System.out.println("Attempt to overload ship! WARNING!!!");
                    break;
                }

                if (dockedShip.getFreeCapacity() == 0) {
                    super.undock();
                }
            }

            try {
                Thread.sleep(timeForOnePiece);
            } catch (InterruptedException e) {
                break;
            }
        }
    }
}

File: study/src/kt/proxy_k/Mine.kt
package kt.proxy_k

import Patterns.proxy.ShipOverloadedException
import Patterns.proxy.SpaceShip

class Mine (x: Double,
            y: Double,
            radius: Double,
            r: Int,
            g: Int,
            b: Int,
            private var warehouseCapacity: Int = 1,
            private var warehouseLoad: Int = 0,
            private var timeForOnePiece: Long = 1000,
            private var maxBatchSize: Int = 2
            ): Planet(x, y, radius, r, g, b), Runnable {

    override fun run() {
        while (!Thread.currentThread().isInterrupted) {
            if (warehouseLoad < warehouseCapacity) {
                warehouseLoad += 1
                println("Warehouse load: $warehouseLoad")
            }

            var flag = true

            dockedShip?.let {
                    var batchSize = maxBatchSize
                    if (it.getFreeCapacity() < maxBatchSize) {
                        batchSize = it.getFreeCapacity()
                    }
                    try {
                        println("Docked ship free capacity: ${it.getFreeCapacity()}")
                        println("Batch size: $batchSize")
                        println("Ship load: ${it.getLoad()}")
                        it.batchLoad(batchSize)
                    } catch (e: ShipOverloadedException) {
                        println("Attempt to overload ship! WARNING!!!")
                        flag = false
                    }

                    if (it.getFreeCapacity() == 0) undock()
            }

            if (!flag) break

            try {
                Thread.sleep(timeForOnePiece)
            } catch (e: InterruptedException) {
                break
            }
        }
    }
}